### PR TITLE
Routing metadata and fixed MAB case study

### LINKED ISSUE #1823: routing metadata missing in new executor
Reporter on community slack by user.

As new orchestrator (executor) is not inspecting payloads passed around the graph it is not currently adding the `routing` metadata, e.g.
```json
"meta" : {
    "puid" : "...",
    "tags" : {},
    "routing" : {
        "ab_test" : 1,
        "combiner1" : -1,
        "mypostprocess1" : -1,
        "router1" : 1
    }
}
```

This may be important in context of multi-armed bandits and AB testing.

### PR DESCRIPTION
**Pull request by axsaucedo**
Fixes #1823 

PR includes:
* Backwards compatibility on routing metadata via predictor process
* Fix of feedback routing logic to read from meta instead of sending route request (aligned to [how engine does it](https://github.com/SeldonIO/seldon-core/blob/7857229949560c7d83bf7eb002beafd46f92ac69/engine/src/main/java/io/seldon/engine/predictors/PredictiveUnitBean.java#L220))
* Updated sklearn / xgboost versions in examples

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **axsaucedo**
You can assign the PR to them by writing `/assign @axsaucedo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Mon Nov 30 15:30:15 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=2

**Comment by seldondev** [comment]
Mon Nov 30 15:30:23 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=1

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 2a8591df147b1c1316665ac5feeee0dc465f55c8 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Tue Dec  1 08:09:55 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=4

**Comment by seldondev** [comment]
Tue Dec  1 08:10:00 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=3

**Comment by seldondev** [comment]
Tue Dec  1 11:54:25 UTC 2020
The logs for [pr-build] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=5

**Comment by seldondev** [comment]
Tue Dec  1 11:54:27 UTC 2020
The logs for [lint] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=6

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 9fde5baae668a645d0c0d17d8fc77b900543415c | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: executor/api/util/utils.go
Comment (cliveseldon) [inline executor/api/util/utils.go:102]: I thought protos were read only - so interesting this works.

**Comment by seldondev** [comment]
Tue Dec  1 13:34:53 UTC 2020
The logs for [pr-build] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=7

**Comment by seldondev** [comment]
Tue Dec  1 13:35:03 UTC 2020
The logs for [lint] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=8

**Comment by seldondev** [comment]
@axsaucedo: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 172b512633bfd66f1042c6c96e83efc1debde17d | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Tue Dec  1 14:16:03 UTC 2020
The logs for [lint] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=10

**Comment by seldondev** [comment]
Tue Dec  1 14:16:10 UTC 2020
The logs for [pr-build] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=9

**Comment by seldondev** [comment]
@axsaucedo: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 172b512633bfd66f1042c6c96e83efc1debde17d | [link]() | `/test this`
lint | 7ab8f6f211be77b6d8dbe5d4e955a277286930c8 | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@axsaucedo: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint | 7ab8f6f211be77b6d8dbe5d4e955a277286930c8 | [link]() | `/test lint`
pr-build | 7ab8f6f211be77b6d8dbe5d4e955a277286930c8 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Tue Dec  1 14:19:50 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=12

**Comment by seldondev** [comment]
Tue Dec  1 14:19:58 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=11

File: executor/api/util/utils.go
Comment (axsaucedo) [reply executor/api/util/utils.go]: Oh ok, I didn't know this, it does seem to work from the tests, so I can double check but all seems good on this imlpementation

**Comment by seldondev** [comment]
Tue Dec  1 14:30:20 UTC 2020
The logs for [pr-build] [13] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/13.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=13

**Comment by seldondev** [comment]
Tue Dec  1 14:30:33 UTC 2020
The logs for [lint] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=14

**Comment by seldondev** [comment]
Tue Dec  1 15:09:08 UTC 2020
The logs for [pr-build] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=15

**Comment by seldondev** [comment]
Tue Dec  1 15:09:14 UTC 2020
The logs for [lint] [16] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/16.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=16

File: executor/predictor/predictor_process.go
Comment (RafalSkolasinski) [inline executor/predictor/predictor_process.go:151]: Separate method because in in case of feedback we route it according to information sent by user?
I guess it's worth to add comment about it, as it's not that obvious at first look.

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]
Nice one! I added few comments / questions.

File: executor/predictor/predictor_process.go
Comment (axsaucedo) [reply executor/predictor/predictor_process.go:151]: Oh I thought it was, this doesn't call the client underneath at all, so it doesn't send any requests.

File: executor/api/util/utils_test.go
Comment (axsaucedo) [reply executor/api/util/utils_test.go:114]: It wouldn't have the route information

File: executor/api/util/utils.go
Comment (axsaucedo) [reply executor/api/util/utils.go:58]: Yeah, although this function existed before, I moved it to utils

File: components/routers/thompson-sampling/ThompsonSampling.py
Comment (axsaucedo) [reply components/routers/thompson-sampling/ThompsonSampling.py:98]: This is just reformatted (although I do agree, but it seems the whoel file has to be updated)

File: executor/api/util/utils.go
Comment (RafalSkolasinski) [reply executor/api/util/utils.go:58]: I was more asking to make sure I understand what's going on. We could incldue some docstring.

File: executor/api/util/utils_test.go
Comment (axsaucedo) [reply executor/api/util/utils_test.go:110]: Oh good shout, fixed

File: components/routers/thompson-sampling/ThompsonSampling.py
Comment (axsaucedo) [reply components/routers/thompson-sampling/ThompsonSampling.py:48]: Same as below

File: components/routers/case_study/models/rf_model/RFModel.py
Comment (axsaucedo) [reply components/routers/case_study/models/rf_model/RFModel.py:49]: Agree (but again reformatted)

**Comment by seldondev** [comment]
Tue Dec  1 18:26:01 UTC 2020
The logs for [pr-build] [17] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/17.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=17

**Comment by seldondev** [comment]
Tue Dec  1 18:26:03 UTC 2020
The logs for [lint] [18] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/18.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=18

File: components/routers/case_study/models/rf_model/RFModel.py
Comment (RafalSkolasinski) [reply components/routers/case_study/models/rf_model/RFModel.py:49]: opened https://github.com/SeldonIO/seldon-core/issues/2728

File: components/routers/epsilon-greedy/EpsilonGreedy.py
Comment (RafalSkolasinski) [reply components/routers/epsilon-greedy/EpsilonGreedy.py:91]: opened https://github.com/SeldonIO/seldon-core/issues/2729

**Review by RafalSkolasinski — APPROVED** [review]
Opened issues for minor follow ups. I think it's good to go!
/approve
/lgtm

**Comment by axsaucedo** [comment]
Seems the approve didn't get triggered
/approve

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2723#issuecomment-737172560 "Approved")*, *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2723#pullrequestreview-542630722 "LGTM")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski,axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2723#issuecomment-737172560 "Approved")*, *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/2723#pullrequestreview-542630722 "LGTM")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com#issuecomment-737172560/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski,axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Wed Dec  2 11:31:53 UTC 2020
The logs for [pr-build] [19] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/19.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=19

**Comment by seldondev** [comment]
Wed Dec  2 11:31:54 UTC 2020
The logs for [lint] [20] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2723/20.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2723 --build=20

**Pull request merged**
